### DISCUSSION TITLE
Comments on API

### DISCUSSION THREAD

**Discussion by Emoun** [chiselverify/chiselverify#30] [Ideas]
These are my comments on the various parts of the API:

### Coverage:

* https://github.com/chiselverify/chiselverify/wiki/Functional-Coverage
Coverpoints, Bins, Condition take string as first argument. Are they really needed?
If they are just for reporting, make them optional and use file/line numbers when no name is given.

* Regarding reporting:
How does this play out compared to other reporting tools?
Could this be more or less easily integrated into the IDE reporting plugin-thing.
Usefull to have a future mindset of this working like a normal coverage tool that an IDE can use.

* Coverage idea:
Right now we have to define coverage points etc.
Maybe a future project records everything and a query can then used to extract the wanted coverage points from it.

* Bins: https://dl.acm.org/doi/abs/10.1145/351240.351266
Isn't an unconditional bit just a conditional bit where the condition is the range?
This is clearly correct, since conditional bins sometimes don't need a range.
The range is theref	ore just syntax sugar for a specific class of conditions.

* Bins:
Can you have a bin without a condition?
If not, the "Condition" object is superfluous, should simply be part of the bin API

* CoverCondition:
I have a feeling its functionality could be unified with CoverPoint after simplifying the Bins.

* Cross*:
Should be able to unify with normal Covers and Bins

### CRV:

* https://github.com/chiselverify/chiselverify/wiki/Constrained-Random-Verification#example
I really don't like "ElseC". What other options are there?
Can't you use the usual if/else constructs to do the same?

* Is it not possible to check whether the constraints are solvable at compile time?
So that the user doesn't have to check?

* Rand vs Randc
Is the repeatability the only difference? then they should be the same object with a (potentially default) parameter.

* Rand
Again string parameters should be optional.

* Constraint operators:
Why can't operator overloading be use instead of prefixing everything with #

* Enable/disable constraints
I dont like it. Should use inheritance instead.
Also avoids having to use variables for constraints, which I don't like too

* How does this compare to quickcheck arbitrary/shrink
How was quickchecking considered?
https://dl.acm.org/doi/abs/10.1145/351240.351266

@schoeberl, @Dobios, and I discussed this at a meeting today.

**Comment by dobios** [comment]
>Regarding reporting:
How does this play out compared to other reporting tools?
Could this be more or less easily integrated into the IDE reporting plugin-thing.
Useful to have a future mindset of this working like a normal coverage tool that an IDE can use.  
  
So after having looked into this, I've found that there isn't a real standard for reporting coverage inside of an IDE. IntelliJ has an [in-depth discussion on how to implement custom-reporting in their IDE](https://intellij-support.jetbrains.com/hc/en-us/community/posts/206103879-Graphical-integration-of-running-tests-in-plugin?page=2#comments), however it is only supported for Java projects for some reason and doesn't work for Scala or Kotlin yet from what I understood. So I don't think that this path is worth exploring for now and should maybe be left for future work...

**Comment by dobios** [comment]
The other comments related to Functional Coverage are solved in PR #31.

**Comment by dobios** [comment]
> * Rand vs Randc
>   Is the repeatability the only difference? then they should be the same object with a (potentially default) parameter.
> 
> * Rand
>   Again string parameters should be optional.
> 
> * Constraint operators:
>   Why can't operator overloading be use instead of prefixing everything with #  
  
These are solved in PR #32.

**Reply by hansemandse** [reply]
I like these updates! Especially that everything is collected into one big function. But I still think the `#`-notation is a bit strange.

**Reply by dobios** [reply]
I agree, however it's only in the case of Random bundles (an experimental feature) where that notation is used, since it essentially converts UInt ports into random variables. We can try to discuss a new notation for it. Overriding the basic operators causes the compiler to not really understand what's going on (it gives out an ambiguous overload error). So we will have to stick with it for now (again only for the Random bundles, so I don't think that it's that much of an issue for the time being).

**Reply by hansemandse** [reply]
True. Let's leave it as is instead of coming up with any more notations.

**Comment by dobios** [comment]
>  Coverage idea:
>       Right now we have to define coverage points etc.
>       Maybe a future project records everything and a query can then used to extract the wanted coverage points from it.

A first draft of this can be found in PR #33.

**Reply by hansemandse** [reply]
Looks good!
